### ISSUE TITLE
Apply a recursion counter to the recent build_display filter recursion limiter

### ISSUE THREAD

**Issue by deanmarcussen** [OrchardCMS/OrchardCore#7451]
Feature added here https://github.com/OrchardCMS/OrchardCore/pull/7357

Need to add counter to it to allow shapes to be built more than once (think bag items, in different display modes).

But still limit the amount of times this can happen to prevent stackoverflows when accidentally called recursively from within the same content template.

**Comment by jeffolmstead** [comment]
As commented on Gitter, my breaking example is:

```
{% for blogPost in recentBlogPosts %}
  {{ blogPost | shape_build_display: "Featured" | shape_render }}
  {{ blogPost | shape_build_display: "Thumbnail" | shape_render }} 
{% endfor %}
```

Thanks for tackling this.

**Comment by sebastienros** [comment]
how is Jeff's bug related to full text? I assumed the recursion limit was only applied to full text generation which is recursive. Unless it has also been implemented to shape display, in which case the key should take the display type into account. But I assume a counter works too, as long as it's a setting.

**Comment by deanmarcussen** [comment]
It slipped in because I tested the other content filters at the same time, and discovered the issue with the `build_display` filter.

I used a counter rather than a display type, as there's been times (in Razor) I've built the display type twice, and then mutated the shapes to display differently (breadcrumbs, tiles, and list items).

So I set it at 20 - which feels really high, but this was about preventing accidental mistakes bringing the site down. It quickly becomes obvious that a mistake in the template has been made.

Could be an Option so it's configurable.
